docs/decisions/0006-pluggable-lti-configuration.rst

```diff
@@ -0,0 +1,121 @@
+Pluggable and re-usable LTI configuration
+-----------------------------------------
+
+
+Status
+======
+
+Proposal
+
+Context
+=======
+
+LTI Tools provide extra funcionality and allows course creators to embed external interactive content
+in the LMS through a standardized setup.
+
+Currently in the Open edX platform:
+
+#. Each LTI tool needs to be configured individually, sometimes requiring multiple valiables to be set to get
+a working integration. This makes the course creation process cumbersome on courses that heavily rely on LTI 
+components. 
+#. There's no support for instance-wide LTI tool configuration or any sort of pluggable LTI configurations.
+
+
+This proposal aims to solve these issues by introducing a mechanism to allow LTI configuration pluggability 
+and changes to the XBlock to allow re-using configurations to reduce instructor friction.
+
+
+Decision
+========
+
+The decision is to enable LTI reusability and pluggability using a filter hook (as defined in `OEP-50`_) to fetch re-usable LTI configurations
+from external plugins. 
+
+With this implementation the plugins are responsible for storing the LTI configuration and credentials, while the LTI consumer handles displaying the content.
+
+To enable this, the LTI Configuration model needs to be modified to:
+
+#. Allow externally configured tools (a new :code:`CONFIG_EXTERNAL` choice for the configuration storage).
+#. Two new fields to store an unique external tool id and the plugin name.
+#. Changes to the :code:`get_lti_consumer` methods to allow instancing a :code:`LtiConsumer` from externally managed configurations.
+
+The filter will allow the XBlock LTI Consumer to query re-usable and/or plugabble LTI configurations and display them to course 
+creators as configuration options in Studio. 
+The filter response should contain all the necessary information for the XBlock to locate and and use the externally managed LTI configuration.
+
+The inputs for the filter pipeline should be one of the following:
+
+* The course ID, so that each filter is able to evaluate and return only configurations available to that context
+* The unique configuration ID, so the filter returns exactly the configuration required.
```

Comment:
I think we should keep in mind how filters operate. The output of a filter should match the input. i.e. you can't just pass a course_id to a filter. What I mean is, filters will run on a sequence, the output from the first filter function will be passed to the next filter function. 

So, for example, you will need to pass to the first filter the `context_key` and a list of `lti_credentaisl`. This filter will then pass on the same parameters, but will add entries to  "lti_credentials" (or remove them).

You don't need to pass the configuration ID. The filter pipeline can just return all valid results, and you can pick the one that matches the stored id.

Reply:
@xitij2000 
>I think we should keep in mind how filters operate. The output of a filter should match the input. i.e. you can't just pass a course_id to a filter. What I mean is, filters will run on a sequence, the output from the first filter function will be passed to the next filter function.

That makes sense, should I come up with a django object and let the filters modify it and pass it forward. I found the decision about the accumulative approach [here](https://github.com/eduNEXT/openedx-filters/blob/main/docs/decisions/0003-hooks-filter-tooling-pipeline.rst), but had to read the source to see how it actually worked (here: https://github.com/eduNEXT/openedx-filters/blob/main/openedx_filters/tooling.py#L207).

There's two interesting things there: (1) filter metadata, always passed when instancing the filter and (2) the accumulated output. We can just trigger the pipeline with an empty dict input `{}` and the context key as metadata and let the filters fill and modify it to add LTI configurations. Is my interpretation correct?

The filters payload is still in the "draft" state, but it mentions this:
>Although filters have a public promise status, and thus maintainability is an important design goal, filters by definition permit that extension developers manipulate data that will be passed along to the lms or cms processes.
At the same time, using fixed attr classes such as what is used in events is not convenient when trying to have few restrictions on what developers can do.

I understood that can take some liberties with how we wanted the return format to be, as long as the response is a dict and the filter and takes the accumulated output of previous functions.

>You don't need to pass the configuration ID. The filter pipeline can just return all valid results, and you can pick the one that matches the stored id.

:+1:

Reply:
Yes, we can trigger it with an empty dict, have the filters manipulate the data along the way, and return the data we want. 

I think the idea is that filters could come in any order (you can control the order, but as a filter developer you don't know the order), so the output from one filter will go to the next filter, so it should be able to process it. A few extra keys in a dict shouldn't break a step in the pipeline.